src/MobileRedirectManager.php

```diff
@@ -0,0 +1,225 @@
+<?php
+/**
+ * Class MobileRedirect.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP;
+
+use AMP_Options_Manager;
+
+/**
+ * Centralized manager to handle mobile redirection of a page.
+ *
+ * @package AmpProject\AmpWP
+ */
+final class MobileRedirectManager {
+
+	/**
+	 * The name of the cookie that persists the user's preference for viewing the non-AMP version of a page when on mobile.
+	 *
+	 * @var string
+	 */
+	const DISABLED_COOKIE_NAME = 'amp_mobile_redirect_disabled';
+
+	/**
+	 * Query parameter to indicate that the page in question should not be served as AMP.
+	 *
+	 * @var string
+	 */
+	const NO_AMP_QUERY_VAR = 'noamp';
+
+	/**
+	 * Get whether mobile redirection is enabled or not.
+	 *
+	 * @return bool If JS redirection is disabled, only the status of the mobile redirection option is returned.
+	 *              Otherwise, returns true if mobile redirection option is enabled and current request is from a mobile device.
+	 */
+	public static function is_enabled() {
+		$option_enabled = AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT );
+
+		return self::should_redirect_via_js() ? $option_enabled : $option_enabled && self::is_mobile_request();
+	}
+
+	/**
+	 * Determine if the current request is from a mobile device.
+	 *
+	 * @return bool True if current request is from a mobile device, otherwise false.
+	 */
+	public static function is_mobile_request() {
+		/**
+		 * Filters whether the current request is from a mobile device. This is provided as a means to short-circuit
+		 * the normal determination of a mobile request below.
+		 *
+		 * @since 1.6
+		 *
+		 * @param bool $is_mobile Whether the current request is from a mobile device.
+		 */
+		$pre_is_mobile = apply_filters( 'amp_pre_is_mobile', false );
+
+		if ( true === $pre_is_mobile ) {
+			return (bool) $pre_is_mobile;
+		}
+
+		$current_user_agent = $_SERVER['HTTP_USER_AGENT'];
+
+		if ( empty( $current_user_agent ) ) {
+			return false;
+		}
+
+		$is_mobile   = false;
+		$user_agents = self::get_user_agents();
+
+		foreach ( $user_agents as $user_agent ) {
+			if ( false !== strpos( $current_user_agent, $user_agent ) ) {
```

Comment:
We may want to consider `$user_agent` being a regular expression pattern rather than just a substring match.

Reply:
Done in a7a72f83bd42e24824f3f3863e730a06ee9e4779.